crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1935 @@
+       out[i] = tmp[i];
+}
+
+/*-
+ * Group operations
+ * ----------------
+ *
+ * Building on top of the field operations we have the operations on the
+ * elliptic curve group itself. Points on the curve are represented in Jacobian
+ * coordinates */
+
+/*-
+ * point_double calculates 2*(x_in, y_in, z_in)
+ *
+ * The method is taken from:
+ *   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-3.html#doubling-dbl-2001-b
+ *
+ * Outputs can equal corresponding inputs, i.e., x_out == x_in is allowed.
+ * while x_out == y_in is not (maybe this works, but it's not tested). */
+static void
+point_double(felem x_out, felem y_out, felem z_out,
+             const felem x_in, const felem y_in, const felem z_in)
+{
+    widefelem tmp, tmp2;
+    felem delta, gamma, beta, alpha, ftmp, ftmp2;
+
+    felem_assign(ftmp, x_in);
+    felem_assign(ftmp2, x_in);
+
+    /* delta = z^2 */
+    felem_square_reduce(delta, z_in);     /* delta[i] < 2^56 */
+
+    /* gamma = y^2 */
+    felem_square_reduce(gamma, y_in);     /* gamma[i] < 2^56 */
+
+    /* beta = x*gamma */
+    felem_mul_reduce(beta, x_in, gamma);  /* beta[i] < 2^56 */
+
+    /* alpha = 3*(x-delta)*(x+delta) */
+    felem_diff64(ftmp, delta);            /* ftmp[i] < 2^60 + 2^58 + 2^44 */
+    felem_sum64(ftmp2, delta);            /* ftmp2[i] < 2^59 */
+    felem_scalar64(ftmp2, 3);             /* ftmp2[i] < 2^61 */
+    felem_mul_reduce(alpha, ftmp, ftmp2); /* alpha[i] < 2^56 */
+
+    /* x' = alpha^2 - 8*beta */
+    felem_square(tmp, alpha);             /* tmp[i] < 2^115 */
+    felem_assign(ftmp, beta);             /* ftmp[i] < 2^56 */
+    felem_scalar64(ftmp, 8);              /* ftmp[i] < 2^59 */
+    felem_diff_128_64(tmp, ftmp);         /* tmp[i] < 2^115 + 2^64 + 2^48 */
+    felem_reduce(x_out, tmp);             /* x_out[i] < 2^56 */
+
+    /* z' = (y + z)^2 - gamma - delta */
+    felem_sum64(delta, gamma);     /* delta[i] < 2^57 */
+    felem_assign(ftmp, y_in);      /* ftmp[i] < 2^56 */
+    felem_sum64(ftmp, z_in);       /* ftmp[i] < 2^56 */
+    felem_square(tmp, ftmp);       /* tmp[i] < 2^115 */
+    felem_diff_128_64(tmp, delta); /* tmp[i] < 2^115 + 2^64 + 2^48 */
+    felem_reduce(z_out, tmp);      /* z_out[i] < 2^56 */
+
+    /* y' = alpha*(4*beta - x') - 8*gamma^2 */
+    felem_scalar64(beta, 4);       /* beta[i] < 2^58 */
+    felem_diff64(beta, x_out);     /* beta[i] < 2^60 + 2^58 + 2^44 */
+    felem_mul(tmp, alpha, beta);   /* tmp[i] < 2^119 */
+    felem_square(tmp2, gamma);     /* tmp2[i] < 2^115 */
+    felem_scalar128(tmp2, 8);      /* tmp2[i] < 2^118 */
+    felem_diff128(tmp, tmp2);      /* tmp[i] < 2^127 + 2^119 + 2^111 */
+    felem_reduce(y_out, tmp);      /* tmp[i] < 2^56 */
+}
+
+/* copy_conditional copies in to out iff mask is all ones. */
+static void copy_conditional(felem out, const felem in, limb mask)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++) {
+        const limb tmp = mask & (in[i] ^ out[i]);
+        out[i] ^= tmp;
+    }
```

Comment:
Is it worthwhile using our constant time #include for this?
Not a firm request but worth thinking about.

Reply:
There aren't any real fit-for-purpose methods there. For example to use `constant_time_cond_swap_buff` we would need to copy the contents of `in` first. This method occurs in a hot path, so keeping original implementation.